api/queryio/client.go

```diff
@@ -0,0 +1,91 @@
+package queryio
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/brimdata/zed/api"
+	"github.com/brimdata/zed/api/client"
+	"github.com/brimdata/zed/driver"
+	"github.com/brimdata/zed/zbuf"
+	"github.com/brimdata/zed/zio/zngio"
+	"github.com/brimdata/zed/zng"
+	"github.com/brimdata/zed/zson"
+)
+
+const maxBatchSize = 100
+
+func RunClientResponse(ctx context.Context, d driver.Driver, res *client.ReadCloser) (zbuf.ScannerStats, error) {
+	format, err := api.MediaTypeToFormat(res.ContentType)
+	if err != nil {
+		return zbuf.ScannerStats{}, err
+	}
+	if format != "zng" {
+		return zbuf.ScannerStats{}, fmt.Errorf("unsupported format: %s", format)
+	}
+	run := &runner{driver: d}
+	r := NewZNGReader(zngio.NewReader(res, zson.NewContext()))
+	for ctx.Err() == nil {
+		rec, ctrl, err := r.ReadPayload()
+		if err != nil {
+			return run.stats, err
+		}
+		if ctrl != nil {
+			run.handleCtrl(ctrl)
+			continue
+		}
+		if rec != nil {
+			run.Write(rec)
+			continue
+		}
+		return run.stats, run.flush()
+	}
+	return run.stats, ctx.Err()
+}
+
+type runner struct {
+	driver driver.Driver
+	cid    int
+	recs   []*zng.Record
+	stats  zbuf.ScannerStats
+}
+
+func (r *runner) Write(rec *zng.Record) error {
+	r.recs = append(r.recs, rec)
+	if len(r.recs) > maxBatchSize {
+		return r.flush()
+	}
+	return nil
+}
+
+func (r *runner) flush() error {
+	if len(r.recs) > 0 {
+		recs := make([]*zng.Record, len(r.recs))
+		copy(recs, r.recs)
+		r.recs = r.recs[:0]
+		return r.driver.Write(r.cid, zbuf.Array(recs))
+	}
+	return nil
+}
```

Comment:
Maybe simplify by getting rid of `flush` and always handing single-record batches to the driver? We can always get more complicated down the road if the extra allocations become a bottleneck.
```suggestion
func (r *runner) Write(rec *zng.Record) error {
	return r.driver.Write(r.cid, &zbuf.Array{rec})
}
```